README.md
NBD Program
===========

Predicting Purchase Rates in Stationary Markets.

```
NBD Program
Predicting purchase rates in stationary markets

(b): 0.2
(w): 2.1
(t): 12 weeks

Penetration (b)
 1 day 1 week 4 weeks 12 weeks 24 weeks 48 weeks 365 days
 0.005  0.032     0.1      0.2    0.277    0.355    0.364

Purchase frequency (w)
 1 day 1 week 4 weeks 12 weeks 24 weeks 48 weeks 365 days
 1.015  1.106   1.401      2.1    3.036    4.729    5.007

Repat purchase (%)
 1 day 1 week 4 weeks 12 weeks 24 weeks 48 weeks 365 days
 2.966  16.82   40.99   61.646   71.615   78.885   79.597

Purchase frequency of buyers
 1 day 1 week 4 weeks 12 weeks 24 weeks 48 weeks 365 days
 1.028  1.184   1.634    2.548    3.666    5.596    5.908

Purchase frequency of new buyers
 1 day 1 week 4 weeks 12 weeks 24 weeks 48 weeks 365 days
 0.005  0.032   0.113    0.304    0.581    1.128    1.222

Proportion -

            1 day 1 week 4 weeks 12 weeks 24 weeks 48 weeks 365 days
not buying  0.995  0.968   0.900    0.800    0.723    0.645    0.636
buying once 0.005  0.029   0.073    0.106    0.114    0.112    0.111
twice       0.000  0.003   0.018    0.043    0.055    0.060    0.060
3 times     0.000  0.000   0.006    0.022    0.033    0.039    0.040
4 times     0.000  0.000   0.002    0.012    0.021    0.028    0.029
5 times     0.000  0.000   0.001    0.007    0.014    0.021    0.022

Proportion of sales due to those buying -

            1 day 1 week 4 weeks 12 weeks 24 weeks 48 weeks 365 days
buying once 0.970  0.820   0.522    0.252    0.135    0.067    0.061
twice       0.029  0.150   0.262    0.206    0.131    0.071    0.066
3 times     0.001  0.025   0.121    0.155    0.117    0.070    0.065
4 times     0.000  0.004   0.054    0.113    0.101    0.067    0.063
5 times     0.000  0.001   0.024    0.081    0.086    0.063    0.059
```

References
----------

* East, R. and Wright, M. and Vanhuele, M. (2013). Consumer Behaviour: Applications
  in Marketing. SAGE Publications, pp. 72-85.

* Repeat-Buying: Facts,Theory and Applications, 2nd edn. A.S.C. Ehrenberg, 1988, London, Charles
  Griffin, ISBN 0 85264 287 3 Book Review: Repeat-Buying:Facts,Theory and Applications by A.S.C.
  Ehrenberg. Norman Pigden. The Statistician, Vol. 40, No. 3, Special Issue (1991), pp. 349-350.

* Brasini, S., Tassinari, F., & Tassinari, G. (1999). Marketing e pubblicità: metodi di analisi
  statistica. Il mulino, pp. 207-215.


nbdprogram/NEWS

Overview of Changes in nbdprogram 0.2:
======================================

* No user-visible news;
* Internal porting of functions to C code.


What's new in nbdprogram 0.1.4:
===============================

* First public release.
* S3 class "nbdprogram" with print method


nbdprogram/R/REstd.R
# nbdprogram/R/REstd.R
# Copyright (c) 2017 Daniele Medri
# Distributed under the GNU GPL-2 license

.fnrod <- function(i, j, r) {
  if (missing(i))
    stop("Argument 'i' must be present.", call. = FALSE)
  if (missing(j))
    stop("Argument 'j' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(0.0)
  
  out <- .C("fnrod",
            as.numeric(i),
            as.numeric(j),
            r)
  return(out[[3]])
}

.fnbval <- function(n, m, g, k, r) {
  if (missing(n))
    stop("Argument 'n' must be present.", call. = FALSE)
  if (missing(m))
    stop("Argument 'm' must be present.", call. = FALSE)
  if (missing(g))
    stop("Argument 'g' must be present.", call. = FALSE)
  if (missing(k))
    stop("Argument 'k' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(rep(0.0, length(g)))

  out <- .C("fnbval",
            as.numeric(n),
            as.numeric(m),
            as.numeric(g),
            as.numeric(k),
            r)
  return(out[[5]])
}

.fnqval <- function(bval, r) {
  if (missing(bval))
    stop("Argument 'bval' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(rep(0.0, length(bval)))

  out <- .C("fnqval",
            as.numeric(bval),
            r)
  return(out[[2]])
}

.fnwval <- function(n, m, g, k, r) {
  if (missing(n))
    stop("Argument 'n' must be present.", call. = FALSE)
  if (missing(m))
    stop("Argument 'm' must be present.", call. = FALSE)
  if (missing(g))
    stop("Argument 'g' must be present.", call. = FALSE)
  if (missing(k))
    stop("Argument 'k' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(rep(0.0, length(g)))

  out <- .C("fnwval",
            as.numeric(n),
            as.numeric(m),
            as.numeric(g),
            as.numeric(k),
            r)
  return(out[[5]])
}

.fndval <- function(bval, wval, k, r) {
  if (missing(bval))
    stop("Argument 'bval' must be present.", call. = FALSE)
  if (missing(wval))
    stop("Argument 'wval' must be present.", call. = FALSE)
  if (missing(k))
    stop("Argument 'k' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(rep(0.0, length(bval)))

  out <- .C("fndval",
            as.numeric(bval),
            as.numeric(wval),
            as.numeric(k),
            r)
  return(out[[4]])
}

.fnmval <- function(bval, wval, k, r) {
  if (missing(bval))
    stop("Argument 'bval' must be present.", call. = FALSE)
  if (missing(wval))
    stop("Argument 'wval' must be present.", call. = FALSE)
  if (missing(k))
    stop("Argument 'k' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(rep(0.0, length(bval)))

  out <- .C("fnmval",
            as.numeric(bval),
            as.numeric(wval),
            as.numeric(k),
            r)
  return(out[[4]])
}

.fneval <- function(bval, dval, mval, r) {
  if (missing(bval))
    stop("Argument 'bval' must be present.", call. = FALSE)
  if (missing(dval))
    stop("Argument 'dval' must be present.", call. = FALSE)
  if (missing(mval))
    stop("Argument 'mval' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(rep(0.0, length(bval)))

  out <- .C("fneval",
            as.numeric(bval),
            as.numeric(dval),
            as.numeric(mval),
            r)
  return(out[[4]])
}

.fnfval <- function(bval, wval, k, r) {
  if (missing(bval))
    stop("Argument 'bval' must be present.", call. = FALSE)
  if (missing(wval))
    stop("Argument 'wval' must be present.", call. = FALSE)
  if (missing(k))
    stop("Argument 'k' must be present.", call. = FALSE)
  if (missing(r))
    r <- as.numeric(rep(0.0, length(bval)))

  out <- .C("fnfval",
            as.numeric(bval),
            as.numeric(wval),
            as.numeric(k),
            r)
  return(out[[4]])
}

.fnqvalrest <- function(bval, wval, qval, k) {
  if (missing(bval))
    stop("Argument 'bval' must be present.", call. = FALSE)
  if (missing(wval))
    stop("Argument 'wval' must be present.", call. = FALSE)
  if (missing(qval))
    stop("Argument 'qval' must be present.", call. = FALSE)
  if (missing(k))
    stop("Argument 'k' must be present.", call. = FALSE)

  out <- .C("fnqvalrest",
            as.numeric(bval),
            as.numeric(wval),
            as.numeric(qval),
            as.numeric(k))
  return(out[[3]])
}

.fnzval <- function(bval, wval, qval, k, zval) {
  if (missing(bval))
    stop("Argument 'bval' must be present.", call. = FALSE)
  if (missing(wval))
    stop("Argument 'wval' must be present.", call. = FALSE)
  if (missing(qval))
    stop("Argument 'qval' must be present.", call. = FALSE)
  if (missing(k))
    stop("Argument 'k' must be present.", call. = FALSE)
  if (missing(zval))
    stop("Argument 'zval' must be present.", call. = FALSE)

  out <- .C("fnzval",
            as.numeric(bval),
            as.numeric(wval),
            as.numeric(qval),
            as.numeric(zval))
  return(out[[4]])
}


.REstd <- function(np) {

  p <- as.numeric(1.0 - np$b)
  m <- as.numeric(np$b * np$w)
  c <- as.numeric((-1 * m) / log(p))
  k <- as.numeric(0.045)
  z <- as.numeric(0.0)

  np$bval <- matrix(0, ncol = 7)
  np$qval <- matrix(0, nrow = 6, ncol = 7)
  np$wval <- matrix(0, ncol = 7)
  np$mval <- matrix(0, ncol = 7)
  np$dval <- matrix(0, ncol = 7)
  np$zval <- matrix(0, nrow = 5, ncol = 7)
  np$eval <- matrix(0, ncol = 7)
  np$fval <- matrix(0, ncol = 7)
  



  while((.fnrod(z, 5.0) != .fnrod(p, 5.0)) || (.fnrod(z, 3.0) != .fnrod(p, 3.0))) {
    y <- (1.0 + m / k)
    z <- (1.0 / (y^k))
  
    if (np$o < 498) {
      np$o <- np$o + 1
    } else if (np$o >= 700) {
      message("Warning: values outside the estimation range.")
    }
    
    if (z < p) {
      np$x <- (np$x - np$v * np$x)
      k <- (k - np$x)
      if (k <= 0) {
        k <- as.numeric(0.00001)
      }
    } else {
      np$x <- (np$x - np$v * np$x)
      k <- (k + np$x)
    }
  }


  np$bval[1,] <- .fnbval(np$n, m, np$g[1,], k)
  np$qval[1,] <- .fnqval(np$bval[1,])
  np$wval[1,] <- .fnwval(np$n, m, np$g[1,], k)
  np$dval[1,] <- .fndval(np$bval[1,], np$wval[1,], k)
  np$mval[1,] <- .fnmval(np$bval[1,], np$wval[1,], k)  
  np$eval[1,] <- .fneval(np$bval[1,], np$dval[1,], np$mval[1,])
  np$dval[1,] <- .fnrod((np$dval[1,]), (np$l + 1))
  np$fval[1,] <- .fnfval(np$bval[1,], np$wval[1,], k)

  a <- matrix(0, ncol = 7)
  a[1,] <- as.numeric(np$bval[1,]*np$wval[1,]/k)
 
  #np$qval <- matrix(fnqvalrest(np$bval, np$wval, np$qval, k), 5, 7)
  #np$zval <- matrix(fnzval(np$bval, np$wval, np$qval, k, np$zval), 5, 7)

  for (j in seq(1:5)) {
    jj <- (j+1)
    for (i in seq(1:7)) {
      np$qval[jj,i] <- as.numeric((a[i]/(1+a[i]))*(1-(a[i]-k*a[i])/(a[i]*j))*np$qval[j,i])
      np$zval[j,i]  <- as.numeric((j*np$qval[jj,i])/(np$wval[i]*np$bval[i]))
    }
  }
  
  class(np) <- "nbdprogram"
  return(np)
}


nbdprogram/R/utils.R
# nbdprogram/R/utils.R
# Copyright (c) 2017 Daniele Medri
# Distributed under the GNU GPL-2 license

.dateseq <- function(startdate) {
  if (missing(startdate))
    startdate <- as.Date(format(Sys.time(), "%Y-%m-%d"))
  enddate <- as.Date(startdate) + 364  
  r <- seq(as.Date(startdate), as.Date(enddate), by="days")
  
  return(r)
}

.check_enum <- function(x, set) {
  if (missing(x) || missing(set))
    stop("Missing needed arguments.", call. = FALSE)
  if (is.null(x) || is.null(set))
    stop("Arguments must be not null.", call. = FALSE)
  if ( (is.character(x) && is.character(set)) ||
        (is.numeric(x) && is.numeric(set))   ||
        (is.logical(x) && is.logical(set)) ) {
    r <- (x %in% set)
  } else {
    stop("Arguments must be of the same type.", call. = FALSE)
  }
  return(r)
}


nbdprogram/R/nbdprogram.R
# nbdprogram/R/nbdprogram.R
# Copyright (c) 2017 Daniele Medri
# Distributed under the GNU GPL-2 license

nbdprogram <- function(b = 0.20, w = 2.1, t = 4) {
  np <- list()

  # b (penetration)
  if (!is.numeric(b))
    stop("Argument 'b' must be numeric.")
  if (!(b > 0) || !(b < 1))
    stop("Argument 'b' must in the range [0,1].")
  
  # w (frequency)
  if (!is.numeric(w))
    stop("Argument 'w' must be numeric.")
  if (!(w >= 1))
    stop("Argument 'w' must equal or more than 1.")
  
  # t (period of time)
  if (!(t %in% c(1, 2, 3, 4, 5, 6, 7, 8)))
    stop("Argument 't' must be in range [1-8]")


  np$l <- as.numeric(2.0)
  np$v <- as.numeric(0.006)
  np$x <- as.numeric(0.045)
  np$o <- 0


  # b (check)
  np$b <- as.numeric(b)
  if (np$b <= 0.025) np$v <- 0.01

  # w (check)
  np$w <- as.numeric(w)
  if (np$w >= 6.0) np$v <- 0.1

  # t (check)
  np$t <- as.integer(t)

  gv <- c(1, 7, 28, 84, 168, 336, 365)
  np$g <- matrix(gv, ncol = 7)
  np$n <- (1.0 / np$g[np$t])    
  
  np <- .REstd(np)

  return(np)
}


nbdprogram/src/functions.c
/*
 * ndbprogram/src/functions.c
 * Copyright (c) 2017 Daniele Medri
 * Distributed under the GNU GPL-2 license
*/

#include <stdio.h>
#include <math.h>

void fnbval(double *n, double *m, double *g, double *k, double *r) {
  int i;

  for (i = 0; i < 7; i++) {        
    r[i] = (1 - 1 / pow((1 + n[0] * m[0] * (g[i] / k[0] )), k[0]));
  }
}

void fnqval(double *bval, double *r) {

  int i;

  for (i = 0; i < 7; i++) {
    r[i] = (1.0 - bval[i]);
  }
}
/*
void fnqvalrest(double *bval, double *wval, double *qval[], double *k) {

  int i, j, jj;
  double a[7];
  int nr = 5;
  int nc = 7;

  for (j = 0; j < nr; j++) {
    jj = (j + 1);
    for (i = 0; i < nc; i++) {
      a[i] = (bval[i] * wval[i] / k[0]);
      qval[i + jj*nr] = ((a[i] / (1 + a[i])) * (1 - (a[i] - k[0] * a[i]) / (a[i] * j)) * qval[i + j*nr]);
    }
  }
}

void fnzval(double *bval, double *wval, double *qval[], double *zval[]) {

  int i, j, jj;
  int nr = 5;
  int nc = 7;

  for (j = 0; j < nr; j++) {
    jj = (j + 1);
    for (i = 0; i < nc; i++) {
      zval[i + jj*nr] = (j * qval[i + jj*nr]) / (wval[i] * bval[i]);
    }
  }
}
*/
void fnwval(double *n, double *m, double *g, double *k, double *r) {
  int i;

  for (i = 0; i < 7; i++) {        
    r[i] = ((n[0] * m[0] * g[i]) / (1 - 1 / pow((1 + m[0] * n[0] * g[i] / k[0]), k[0])));
  }
}


void fndval(double *bval, double *wval, double *k, double *r)
{
  int i;

  for (i = 0; i < 7; i++) {
    r[i] = (100 / bval[i]) * (1 - 2 / pow((1 + bval[i] * wval[i] / k[0]), k[0]) + 1 / pow((1 + 2 * bval[i] * wval[i] / k[0]), k[0]));
  }
}

void fnmval(double *bval, double *wval, double *k, double *r)
{
  int i;

  for (i = 0; i < 7; i++) {
    r[i] = (wval[i] * bval[i] * (1 - 1 / pow((1 + wval[i] * bval[i] / k[0]), (k[0] + 1))));
  }
}


void fneval(double *bval, double *dval, double *mval, double *r)
{
  int i;

  for (i = 0; i < 7; i++) {
    r[i] = (mval[i] / (dval[i] * bval[i] / 100));
  }
}

void fnfval(double *bval, double *wval, double *k, double *r)
{
  int i;

  for (i = 0; i < 7; i++) {
    r[i] = (pow((1 + bval[i] * wval[i] / k[0]), (k[0] + 1)) * (bval[i] - (1 - 2 * pow((1 + bval[i] * wval[i] / k[0]), (-k[0])) + pow((1 + 2 * bval[i] * wval[i] / k[0]), (-k[0])))));
  }
}


nbdprogram/R/print.R
# nbdprogram/R/print.R
# Copyright (c) 2017 Daniele Medri
# Distributed under the GNU GPL-2 license

print.nbdprogram <- function(x, ...) {

  message("NBD Program")
  message("Predicting purchase rates in stationary markets\n")
  tdesc <- c("1 day", "1 week", "4 weeks", "12 weeks",
               "24 weeks", "48 weeks", "365 days")
  message(paste("(b):",x$b))
  message(paste("(w):",x$w))
  message(paste("(t):",tdesc[x$t]))

  message("\nPenetration (b)")
  p <- as.data.frame(round(x$bval,3), row.names="")
  names(p) <- tdesc
  print(p)

  message("\nPurchase frequency (w)")
  p <- as.data.frame(round(x$wval,3), row.names="")
  names(p) <- tdesc
  print(p)

  message("\nRepat purchase (%)")
  p <- as.data.frame(round(x$dval,3), row.names="")
  names(p) <- tdesc
  print(p)

  message("\nPurchase frequency of buyers")
  p <- as.data.frame(round(x$eval,3), row.names="")
  names(p) <- tdesc
  print(p)

  message("\nPurchase frequency of new buyers")
  p <- as.data.frame(round(x$fval,3), row.names="")
  names(p) <- tdesc
  print(p)
  
  # 2nd table
  
  message("\nProportion -\n")
  rn <- c("not buying", "buying once", "twice", "3 times", "4 times", "5 times")
  p <- as.data.frame(round(x$qval,3), row.names=rn)
  names(p) <- tdesc
  print(p)

  # 3rd table
  
  message("\nProportion of sales due to those buying -\n")
  p <- as.data.frame(round(x$zval,3), row.names=rn[2:6])
  names(p) <- tdesc
  print(p)

}
